### PR TITLE
Release 2.7.0

### PR DESCRIPTION
**Pull request by ravage84**
We are finally ready to release PHPMD 2.7.0! :rocket: 

This release contains [a lot changes](https://github.com/phpmd/phpmd/commits/master?after=dfaa509cb69d6ad796c90b9e084cb311eb26a2fa+244), actually [over 250 commits](https://github.com/phpmd/phpmd/compare/2.6.0...master).
The release, which includes new features, improvements & fixes is long overdue.

It really needs to get out of the window.
Nonetheless, it should not be rushed.
I targeted the actual merge of this and the tag & GH release for tomorrow, 29.07.2019.

 I kindly ask you to review this thoroughly.

For this, I created a check list:

- [x] All relevant pull requests and commits for 2.7.0 are included in changes.xml
- [x] All the 2.7.0 change entries in changes.xml have the correct date (commit hash)
- [x] All the 2.7.0 change entries in changes.xml have the correct developer
- [x] All the 2.7.0 change entries in changes.xml have the correct type of change
- [x] All the 2.7.0 change entries in changes.xml have the correct change system (all github)
- [x] All the 2.7.0 change entries in changes.xml have the correct issue number
- [x] All the 2.7.0 change entries in changes.xml have have a understandable change text
- [x] All the 2.7.0 change entries in changes.xml have the type of due to mention, if relevant

Please, take note that you cannot simply use the [comparison between 2.6.0 and master](https://github.com/phpmd/phpmd/compare/2.6.0...master), as it only shows the 250 newest changes. For the first few commits, you have to use [the commit history starting after the last commit for 2.6.0](https://github.com/phpmd/phpmd/commits/master?after=dfaa509cb69d6ad796c90b9e084cb311eb26a2fa+244).

Please state which checks you performed in your review.

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,3 +1,167 @@
...
+- Fixed #599: Fix Composer package type Fixed in commit #0ca4eff.
+- Fixed #604: Rename mikey179/vfsStream to mikey179/vfsstream to 
+  prevent Composer error Fixed in commit #f66247f.
+- Fixed #609: Fix main logo link in website build script Fixed in 
+  commit #f3b68be.
+- Fixed #572: Allow different letter casing for @suppressWarnings 
+  annotation Fixed in commit #bb2cfe9.
+- Fixed #575: Fix UnusedFormalParameter false positive in string 
+  compound variable Fixed in commit #8790cbb.
+- Fixed #599: Fix Composer package type Fixed in commit #0ca4eff.
```
Comment (kylekatarnls) [inline CHANGELOG]: Duplicated line `Fixed #599`

File: CHANGELOG [hunk 2]
```diff
@@ -1,3 +1,167 @@
...
+- Fixed #626: Convert special chars in violation description for XML 
+  output Fixed in commit #5305f5b.
+- Fixed #631: Fix url to "How to create a custom rule set" 
+  documentation page Fixed in commit #3c5b534.
+- Fixed #632: Use local PHPUnit in Scrutinizer Fixed in commit 
+  #dfaa509.
+- Fixed #633: Fix AppVeyor build Fixed in commit #40189f3.
```
Comment (kylekatarnls) [inline CHANGELOG]: I would swap the order: `Implemented` first (most important for users), then `Fixed` after.

File: CHANGELOG [hunk 3]
```diff
@@ -1,3 +1,167 @@
...
+- Implemented #443: Support for compound variables in 
+  UnusedLocalVariable Implemented in commit #c7009d5.
+- Implemented #448: Use HTTPS instead of HTTP Implemented in commit 
+  #51eb887.
+- Implemented #454: Use HTTPS instead of HTTP for test files 
+  Implemented in commit #f1c1426.
+- Implemented #451: Use HTTPS instead of HTTP for main files 
+  Implemented in commit #9a77c48.
```
Comment (kylekatarnls) [inline CHANGELOG]: Duplication: original commit + merge (merge should be omitted to keep the authorship)

File: CHANGELOG [hunk 4]
```diff
@@ -1,3 +1,167 @@
...
+- Implemented #455: Use HTTPS instead of HTTP for resource files 
+  Implemented in commit #b073ad2.
+- Implemented #381: Add annotations to allow IDE to reference correct 
+  classes Implemented in commit #2dbae11.
```
Comment (kylekatarnls) [inline CHANGELOG]: This is not an `Implemented`: it provides nothing to the user.

File: CHANGELOG [hunk 5]
```diff
@@ -1,3 +1,167 @@
...
+  output Fixed in commit #5305f5b.
+- Fixed #631: Fix url to "How to create a custom rule set" 
+  documentation page Fixed in commit #3c5b534.
+- Fixed #632: Use local PHPUnit in Scrutinizer Fixed in commit 
+  #dfaa509.
+- Fixed #633: Fix AppVeyor build Fixed in commit #40189f3.
+- Implemented #443: Support for compound variables in 
+  UnusedLocalVariable Implemented in commit #c7009d5.
+- Implemented #448: Use HTTPS instead of HTTP Implemented in commit 
+  #51eb887.
```
Comment (kylekatarnls) [inline CHANGELOG]: This is not an `Implemented`: it provides nothing to the user.

File: CHANGELOG [hunk 6]
```diff
@@ -1,3 +1,167 @@
...
+- Implemented #454: Use HTTPS instead of HTTP for test files 
+  Implemented in commit #f1c1426.
```
Comment (kylekatarnls) [inline CHANGELOG]: This is not an `Implemented`: it provides nothing to the user, mostly since it's only in tests files.

File: CHANGELOG [hunk 7]
```diff
@@ -1,3 +1,167 @@
...
+- Implemented #451: Use HTTPS instead of HTTP for main files 
+  Implemented in commit #9a77c48.
+- Implemented #455: Use HTTPS instead of HTTP for resource files 
+  Implemented in commit #b073ad2.
+- Implemented #381: Add annotations to allow IDE to reference correct 
+  classes Implemented in commit #2dbae11.
+- Implemented #458: Updated PHPCS dev dependency from 2.3.4 to 2.8.1 
+  Implemented in commit #f2ae09f.
```
Comment (kylekatarnls) [inline CHANGELOG]: https://github.com/phpmd/phpmd/commit/f2ae09f seems to miss compose.json change

File: CHANGELOG [hunk 8]
```diff
@@ -1,3 +1,167 @@
...
+- Implemented #459: Integrate Stickler CI Implemented in commit 
+  #d106330.
+- Implemented #458: Update PHPCS & add Composer scripts Implemented in 
+  commit #24ff5a9.
```
Comment (kylekatarnls) [inline CHANGELOG]: Duplicated `Implemented #458`

**Comment by tvbeek** [comment]
I agree with @kylekatarnls that the changelog should contain the changes for the user and not the other changes (code style, typos in documentation, ...)

Maybe it is also an idea to have some headers between the different parts:
## Added:
- ...
- ...

## Fixed
 - ...
- ...

**Comment by ravage84** [comment]
Thank you for your feedback. I guess I should have explained myself better. Instead of doing what I personally would prefer to do, I went and did what we have done in the past.

I'm not sure what we included and what we excluded, so I went through all the commits and tried to include (almost) everything. At least I wanted to add each and every PR that went in.

@kylekatarnls 

> `Implement` keyword should be reserved for actual new features of the library.

This is word is auto-generated through the Ant target. Like I said, that's the way it used to be.
Please mostly ignore the change log, focus on the changes.xml file, instead.

> The CHANGELOG should help the user to see what's new. If he want a technical history of the repository, he/she can use the GIT history.

I agree on that in principial. For me there are three or four distinct things:

1. Commit history
2. Issue tracking (including PRs)
3. Change log
4. Release notes

And yes, the change log should not be a 1:1 copy of any of the other two, nor a full combined list.

> Our CHANGELOG and release notes should mention only new features and features changes in the public API (and eventually protected scope).
> 
> That should not include in my opinion:
> 
>     * Code style (if no risk of behavior change)
>     * Settings changes of tooling (Travis, PHPUnit, Scrutinizer)
>     * Documentation or any part that are not in the final downloaded package
>     * Comments (except eventually for PHPDoc of public methods/property/classes)
>     * Typos (change of wording in descriptions, in --help, options explanation)

It should contain a list of things that could be relevant for the maintainers, contributors and/or end developers, whatever it is. And website buildng & documentation changes _could_ be relevant for some of them in some cases.

The release notes can be trimmed down to the most relevant and link to the change log. This way the visibility is the greatest for the most important stuff but people can easily learn more.

> The CHANGELOG should mention initial commits rather than merge commit whenever possible to keep the authorship.

Well, personally, I wouldn't do that in the future, since we merge through PRs, we can simply link to the related PR(s).
I'm not 100 % sure how it was handled in the past if a change was spead throughout several commits, but for me iit's better to see the full change at once. I don't want nor need to follow every commit. The end result is what is actually included in the code I get with a given release.

> Each change should appear only once in the CHANGELOG.

Agreed. If something is mentioned twice, it's just my fault.

> We should add a BC notice, in this release we have minimum > maximum as a break change. Even if it was an obvious mistake, users may have started to use it and will need to correct their settings. A small explanation: "We're aware it's a BC and apologize for the disturbance, we couldn't reasonably upgrade to a major release and had to make a pragmatic choice regarding the renaming of this option."

Great idea. Thanks for bringing this up. I will include it in the GitHub releae text, as it is most visible there. But we can change the current text of the related text to better show what has changed, too.

@tvbeek 

> I agree with @kylekatarnls that the changelog should contain the changes for the user and not the other changes (code style, typos in documentation, ...)

See my explanations above.

> Maybe it is also an idea to have some headers between the different parts:

In the future, I'd like to adopt [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), as previously discussed.

**Comment by ravage84** [comment]
With that explained, please let's review the changes.xml.
If something should be removed because it's duplicated or not relevant enough, tell me.
Le'ts ignore the way the current change log is worded and how it looks. We can improve on that in the future. One step at a time. I want to get this important release out of the window.

File: CHANGELOG (hunk 8 above)
Comment (ravage84) [reply CHANGELOG]: Please tell me which are duplicated in the chagnes.xml, instead. So I can remove it there and regenerate the change log.

File: CHANGELOG (hunk 7 above)
Comment (ravage84) [reply CHANGELOG]: Well, that wasn't necessary. The composer.lock file __was__ there for that purpose. Now it is deleted... If we build now, it should take the newest dependencies by default. I guess, we can remove that line. Please mention which line it is in the changes.xml.

File: CHANGELOG (hunk 6 above)
Comment (ravage84) [reply CHANGELOG]: Like I said, auto-generated. Let's focus on the actual list of changes for this release.

File: CHANGELOG (hunk 5 above)
Comment (ravage84) [reply CHANGELOG]: Like I said, auto-generated. Let's focus on the actual list of changes for this release.

File: CHANGELOG (hunk 4 above)
Comment (ravage84) [reply CHANGELOG]: Like I said, auto-generated. Let's focus on the actual list of changes for this release.

File: CHANGELOG (hunk 3 above)
Comment (ravage84) [reply CHANGELOG]: Please mention the entry in chagnges.xml. Thanks

File: CHANGELOG (hunk 2 above)
Comment (ravage84) [reply CHANGELOG]: OK, please mention that in the changes.xml

File: CHANGELOG (hunk 1 above)
Comment (ravage84) [reply CHANGELOG]: OK, please mention that in the changes.xml

File: CHANGELOG [hunk 9]
```diff
@@ -1,3 +1,167 @@
...
+  non-existing path Fixed in commit #67bd7c6.
+- Fix invalid "array of strings" type hints Fixed in commit #16e4eda.
+- Fixed #480: Start tag expected, '<' not found Fixed in commit 
+  #3e2e058.
+- Fixed #482: Rename minimum property to maximum in 
+  CouplingBetweenObjects Fixed in commit #9210116.
+- Fixed #494: Fix UnusedPrivateField false positive Fixed in commit 
+  #99f3ba9.
+- Fixed #567: Minor fix to Clean Code Rules documentation Fixed in 
+  commit #175b08f.
```
[the remark this answers is not in the document]
Comment (ravage84) [reply CHANGELOG]: OK, please mention that in the changes.xml

**Comment by ravage84** [comment]
Two reasons on why I haven't removed all the "unimportant" changes:

1. This is the culmination of two and a half years of accumulated changes by many contributors. Now, we finally release a version. Let's relfect the fact that there were over 250 commits since the last minor version.
2. Attributing the contributors of these changes in our changelog through a link to their PR can be motivating. Let's mention all these changes, even if small.

By the way in the release notes, I've planned to only mention the most important ones and link to the change log. Interested parties can follow that link.

S, please, let's review this and get it flying.

**Comment by tvbeek** [comment]
It is a big list to review, so for me it will take some time.
Did you generate the changes.xml with a tool or is it a manual job?

**Comment by ravage84** [comment]
> It is a big list to review, so for me it will take some time.

Sure, take your time. A day or two more won't hurt after 2 and a half years :smirk_cat: 

> Did you generate the changes.xml with a tool or is it a manual job?

Nope, it's "hand-made" but the CHANGELOG is generated based on the changes.xml by a Ant target.
So, you can mostly ignore that.

File: src/site/docx/changes.xml [hunk 10]
```diff
@@ -10,6 +10,250 @@
...
+                 It contains all the new features, improvements and fixes from two and a half years since 2.6.0.
+                 Please take note of a backwards compatibility breaking property renaming in the CouplingBetweenObjects rule.
+                 With this release we also drop support for PHP 5.3.">
+            <action date="716ecf5" dev="eeree" type="add" system="github" issue="472">
+                Add rule for assignment within conditional (IfStatementAssignment)
+            </action>
+            <action date="0e30d82" dev="eeree" type="add" system="github" issue="490">
+                Add rule for count in loop (CountInLoop)
+            </action>
+            <action date="a295850" dev="eeree" due-to="rafalwrzeszcz" type="add" system="github" issue="484">
```
Comment (kylekatarnls) [inline src/site/docx/changes.xml]: Just a question, do you know where/how `dev="eeree"` and `due-to="rafalwrzeszcz"` will appear?
https://github.com/phpmd/phpmd/pull/484

Comment (ravage84) [reply src/site/docx/changes.xml]: To be honest, I do not know. I thought the old website. But apparently not:
https://web.archive.org/web/20190201050403/https://phpmd.org/

I guess, you have to ask a Java developer who knows the Maven toolchain better. I think the current setup tools are inspired by that.

Comment (kylekatarnls) [reply src/site/docx/changes.xml]: OK, we will improve this later to make it appear automatically on the website. In the particular case above rafalwrzeszcz sent commits, other `due-to` in the 2.7.0 CHANGELOG are issue posters.

File: src/site/docx/changes.xml [hunk 11]
```diff
@@ -10,6 +10,250 @@
...
+                Added new options to CLI (min-priority, minimum-priority, report-file, input-file, not-strict)
+            </action>
+            <action date="45de3be" dev="tominventisbe" type="add" system="github" issue="579">
+                Added support for setting a maximum execution priority through CLI (max-priority, maximum-priority, maximumpriority)
+            </action>
+            <action date="63047d9" dev="eeree" due-to="schinkel" type="add" system="github" issue="489">
+                Added new predefined variables to AbstractLocalVariable rule
+            </action>
+            <action date="609c6bb" dev="dxops" type="update" system="github" issue="382">
+                Ignored isser-, hasser-, wither-method for TooManyMethods
```
Comment (ravage84) [inline src/site/docx/changes.xml]: "Ignore" The rule now always ignores it. If you want:
"Made TooManyMethods rule to ignore isser-, hasser-, wither-methods"

File: src/site/docx/changes.xml [hunk 12]
```diff
@@ -10,6 +10,250 @@
...
+            </action>
+            <action date="43d4ed0" dev="ravage84" type="update" system="github" issue="625">
+                Skipped DuplicatedArrayKey rule check, if there is no array key...
```
Comment (ravage84) [inline src/site/docx/changes.xml]: "Skip". Same here.
"Made DuplicatedArrayKey rule to skip, if there is no array key".

File: src/site/docx/changes.xml [hunk 13]
```diff
@@ -10,6 +10,250 @@
...
+                Changed Scrutinizer settings to use PHPUnit
+            </action>
+            <action date="4514235" dev="ravage84" type="add" system="github" issue="460">
+                Added ApiGen config file
+            </action>
+            <action date="2b55442" dev="eeree" type="update" system="github" issue="471">
+                Modified default PHP installation directory to match Chocolatey package in AppVeyor config
+            </action>
+            <action date="252b178" dev="OndraM" type="update" system="github" issue="552">
+                Update PHP in AppVeyor builds to 7.1
```
Comment (ravage84) [inline src/site/docx/changes.xml]: Updated

File: src/site/docx/changes.xml [hunk 14]
```diff
@@ -10,6 +10,250 @@
...
+                Added PHPMD Gitter Community Channel notifications for Travis
+            </action>
+            <action date="d106330" dev="ravage84" type="add" system="github" issue="459">
+                Integrated Stickler CI
+            </action>
+            <action date="9b18153" dev="ravage84" type="update" system="github" issue="492">
+                Adjusted Stickler-CI config for ignoring test resource files
+            </action>
+            <action date="dfaa509" dev="kylekatarnls" type="fix" system="github" issue="632">
+                Changed Scrutinizer settings to use PHPUnit
```
Comment (ravage84) [inline src/site/docx/changes.xml]: Is there a reason you removed the word "local"?

Comment (kylekatarnls) [reply src/site/docx/changes.xml]: https://github.com/phpmd/phpmd/pull/642/commits/7314cd37feca94a1979f99a1714fc447d4f1035b

File: src/site/docx/changes.xml (hunk 13 above)
Comment (kylekatarnls) [reply src/site/docx/changes.xml]: https://github.com/phpmd/phpmd/pull/642/commits/7314cd37feca94a1979f99a1714fc447d4f1035b

File: src/site/docx/changes.xml (hunk 12 above)
Comment (kylekatarnls) [reply src/site/docx/changes.xml]: https://github.com/phpmd/phpmd/pull/642/commits/7314cd37feca94a1979f99a1714fc447d4f1035b

File: src/site/docx/changes.xml (hunk 11 above)
Comment (kylekatarnls) [reply src/site/docx/changes.xml]: https://github.com/phpmd/phpmd/pull/642/commits/7314cd37feca94a1979f99a1714fc447d4f1035b

**Comment by ravage84** [comment]
Now, this needs to wait for https://github.com/phpmd/phpmd/pull/643 and then we should include the change entry just below the original bump change.
Once that is done, it just needs two approvals and we are good to go.

**Pull request merged**